obspy/signal/polarization.py

```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
     dplan = signal.lfilter(fk, 1, plan_add)
     dplan = dplan[len(fk) - 1:]
-    # dplan = dplan[np.size(fk) // 2:(np.size(dplan) - np.size(fk) // 2)]
 
     return leigenv1, leigenv2, leigenv3, rect, plan, dleigenv, drect, dplan
+
+
+def flinn(stream, noise_thres=0):
+    """
+    Computes the azimuth, incidence, rectilinearity and planarity after the
+    eigenstructure decomposition method of [Flinn1965b]_.
+
+    :param stream: ZNE sorted trace data
+    :type stream: :class:`~obspy.core.stream.Stream`
+    :param noise_tresh: Variance of the noise sphere; data points are excluded
+        when falling within the sphere with radius sqrt(noise_thres),
+        default is set to 0.
+    :type noise_thres: float
+    :returns:  azimuth, incidence, rectilinearity, and planarity
+    """
+    Z = []
+    N = []
+    E = []
+    comp, npts = np.shape(stream)
+
+    for i in range(0, npts):
+        if (stream[0][i] ** 2 + stream[1][i] ** 2 + stream[2][i] ** 2) \
+                > noise_thres:
+            Z.append(stream[0][i])
+            N.append(stream[1][i])
+            E.append(stream[2][i])
+
+    X = np.vstack((E, N, Z))
+    covmat = np.cov(X)
+    eigvec, eigenval, v = (np.linalg.svd(covmat))
+    # Rectilinearity defined after Montalbetti & Kanasewich, 1970
+    rect = 1.0 - np.sqrt(eigenval[1] / eigenval[0])
+    # Planarity defined after Jurkevics, 1988
+    plan = 1.0 - (2.0 * eigenval[2] / (eigenval[1] + eigenval[0]))
+    azimuth = math.degrees(math.atan2(eigvec[0][0], eigvec[1][0]))
+    eve = np.sqrt(eigvec[0][0] ** 2 + eigvec[1][0] ** 2)
+    incidence = math.degrees(math.atan2(eve, eigvec[2][0]))
+    if azimuth < 0.0:
+        azimuth = 360.0 + azimuth
+    if incidence < 0.0:
+        incidence += 180.0
+    if incidence > 90.0:
+        incidence = 180.0 - incidence
+        if azimuth > 180.0:
+            azimuth -= 180.0
+        else:
+            azimuth += 180.0
+    if azimuth > 180.0:
+        azimuth -= 180.0
+
+    return azimuth, incidence, rect, plan
+
+
+def instantaneous_frequency(data, sampling_rate):
+    """
+    Simple function to estimate the instantaneous frequency based on the
+    derivative of the data and the analytical (hilbert) data.
+
+    :param data: The data array.
+    :type data: :class:`numpy.ndarray`
+    :param sampling_rate: The sampling rate in Hz.
+    :type sampling_rate: float
+    """
+    x = data.copy()
+    X = signal.hilbert(x)
+    DX = np.gradient(X) * sampling_rate
+
+    instf = (X.real * DX.imag - X.imag * DX.real) / \
+            (2 * math.pi * (abs(X) ** 2))
+
+    return instf
+
+
+def vidale_adapt(stream, noise_thres, fs, flow, fhigh, spoint, stime, etime):
+    """
+    Adaptive window polarization analysis after [Vidale1986]_ with the
+    modification of adapted analysis window estimated by estimating the
+    instantaneous frequency. It returns the azimuth, incidence, rectilinearity
+    planarity and ellipticity.
+
+    :param stream: ZNE sorted trace data
```

Comment:
Why does it need to be sorted? Isn't that what the `stream.sort` call does below?